src/emulator/auth/operations.ts

```diff
@@ -1597,6 +1594,136 @@ function listVerificationCodesInProject(
+  // autoRetrievalInfo is ignored because SMS will not actually be sent.
+
+  const phoneNumber = reqBody.phoneEnrollmentInfo.phoneNumber;
+  assert(phoneNumber, "((Missing phoneNumber.))");
+  assert(isValidPhoneNumber(phoneNumber), "((Invalid phone number.)");
+
+  const { sessionInfo, code } = state.createVerificationCode(phoneNumber);
+
+  // Print out a developer-friendly log containing the link, in lieu of sending
+  // a real text message out to the phone number.
+  EmulatorLogger.forEmulator(Emulators.AUTH).log(
+    "BULLET",
+    `To enroll MFA with ${phoneNumber}, use the code ${code}.`
+  );
+
+  return {
+    phoneSessionInfo: {
+      sessionInfo,
+    },
+  };
+}
+
+function mfaEnrollmentFinalize(
+  state: ProjectState,
+  reqBody: Schemas["GoogleCloudIdentitytoolkitV2FinalizeMfaEnrollmentRequest"]
+): Schemas["GoogleCloudIdentitytoolkitV2FinalizeMfaEnrollmentResponse"] {
+  assert(reqBody.idToken, "MISSING_ID_TOKEN");
+  const { user, signInProvider } = parseIdToken(state, reqBody.idToken);
+  assert(reqBody.phoneVerificationInfo, "((Missing phoneVerificationInfo.))");
+
+  if (reqBody.phoneVerificationInfo.androidVerificationProof) {
+    throw new NotImplementedError("androidVerificationProof is unsupported!");
+  }
+  const { code, sessionInfo } = reqBody.phoneVerificationInfo;
+
+  assert(code, "((Missing phoneVerificationInfo.code.))");
+  assert(sessionInfo, "((Missing phoneVerificationInfo.sessionInfo.))");
+
+  const phoneNumber = verifyPhoneNumber(state, sessionInfo, code);
+
+  const existingFactors = user.mfaInfo || [];
+  const existingIds = new Set<string>();
+  for (const { mfaEnrollmentId } of existingFactors) {
+    if (mfaEnrollmentId) {
+      existingIds.add(mfaEnrollmentId);
+    }
+  }
+  state.updateUserByLocalId(user.localId, {
+    mfaInfo: [
+      ...existingFactors,
+      {
+        displayName: reqBody.displayName,
+        enrolledAt: new Date().toISOString(),
+        mfaEnrollmentId: newRandomId(28, existingIds),
+        phoneInfo: phoneNumber,
+        unobfuscatedPhoneInfo: phoneNumber,
+      },
+    ],
+  });
+
+  const { idToken, refreshToken } = issueTokens(state, user, signInProvider);
+  return {
+    idToken,
+    refreshToken,
+  };
+}
+
+function mfaSignInStart(
+  state: ProjectState,
+  reqBody: Schemas["GoogleCloudIdentitytoolkitV2StartMfaSignInRequest"]
+): Schemas["GoogleCloudIdentitytoolkitV2StartMfaSignInResponse"] {
+  assert(reqBody.mfaPendingCredential, "((Missing mfaPendingCredential.))");
+  assert(reqBody.mfaEnrollmentId, "((Missing mfaEnrollmentId.))");
+  const { user } = parsePendingCredential(state, reqBody.mfaPendingCredential);
+
+  const enrollment = user.mfaInfo?.find(
+    (factor) => factor.mfaEnrollmentId === reqBody.mfaEnrollmentId
+  );
+  assert(enrollment, "((mfaEnrollmentId not found.))");
+  const phoneNumber = enrollment.unobfuscatedPhoneInfo;
+  assert(phoneNumber, "((not a phone-based MFA enrollment.))");
```

Comment:
Nit on capitalization
```suggestion
  assert(phoneNumber, "((Not a phone-based MFA enrollment.))");
```